yktrace/src/sir.rs

```diff
@@ -1,92 +1,148 @@
 //! Loading and tracing of Serialised Intermediate Representation (SIR).
 
 use fallible_iterator::FallibleIterator;
+use fxhash::FxHashMap;
 use memmap::Mmap;
 use object::{Object, ObjectSection};
 use std::{
-    collections::HashMap,
     convert::TryFrom,
     env,
     fmt::{self, Debug, Display, Write},
     fs::File,
-    io::Cursor,
-    iter::Iterator,
-    path::Path
+    io::{Cursor, Seek, SeekFrom},
+    iter::Iterator
 };
-use ykpack::{self, Body, BodyFlags, CguHash, Decoder, Pack, Ty};
+use ykpack::{self, Body, BodyFlags, CguHash, Decoder, Pack, SirHeader, SirOffset};
 
-/// The serialised IR loaded in from disk. One of these structures is generated in the above
-/// `lazy_static` and is shared immutably for all threads.
-#[derive(Debug)]
-pub struct Sir {
-    /// Lets us map a symbol name to a SIR body.
-    pub bodies: HashMap<String, Body>,
-    /// SIR Local variable types, keyed by codegen unit hash.
-    pub types: HashMap<CguHash, Vec<Ty>>
-}
-
-impl Sir {
-    pub fn ty(&self, id: &ykpack::TypeId) -> &ykpack::Ty {
-        &self.types[&id.0][usize::try_from(id.1).unwrap()]
-    }
+lazy_static! {
+    pub static ref EXE_MMAP: Mmap =
+        unsafe { Mmap::map(&File::open(&env::current_exe().unwrap()).unwrap()).unwrap() };
+    pub static ref SIR: Sir<'static> = Sir::new(&*EXE_MMAP).unwrap();
 }
 
-lazy_static! {
-    pub static ref SIR: Sir = Sir::read_file(&env::current_exe().unwrap()).unwrap();
+/// An interface to the serialised IR of an executable.
+///
+/// One of these structures is generated in the above `lazy_static` and is then shared immutably
+/// accross all threads. Only the headers of each SIR section are eagerly loaded. For performance
+/// reasons, the actual IR is loaded lazily on-demand.
+#[derive(Debug)]
+pub struct Sir<'m> {
+    /// The SIR section headers.
+    /// Maps a codegen unit hash to a `(section-name, header, header-size)` tuple.
+    hdrs: FxHashMap<CguHash, (String, SirHeader, SirOffset)>,
+    /// The current executable's ELF information.
+    exe_obj: object::File<'m>
 }
 
-impl Sir {
-    pub fn read_file(file: &Path) -> Result<Sir, ()> {
+impl<'m> Sir<'m> {
+    pub fn new(mmap: &'m Mmap) -> Result<Self, ()> {
         // SAFETY: Not really, we hope that nobody changes the file underneath our feet.
-        let data = unsafe { Mmap::map(&File::open(file).unwrap()).unwrap() };
-        let object = object::File::parse(&*data).unwrap();
-
-        // We iterate over ELF sections, looking for ones which contain SIR and loading them into
-        // memory.
-        let mut bodies = HashMap::new();
-        let mut types = HashMap::new();
-        for sec in object.sections() {
+        fn read_hdr(curs: &mut Cursor<&[u8]>) -> SirHeader {
```

Comment:
Why is this a nested function? It only seems to be called once?

Reply:
I felt it looked nicer. I can unroll it.

Reply:
Putting it in a function made me hunt around to see why it was being called multiple times, so I think it's a distraction.